google-cloud-pubsublite/src/main/java/com/google/cloud/pubsublite/cloudpubsub/internal/AckSetTrackerImpl.java

```diff
@@ -36,13 +36,50 @@
 import java.util.concurrent.atomic.AtomicBoolean;
 
 public class AckSetTrackerImpl extends TrivialProxyService implements AckSetTracker {
+  // Receipt represents an unacked message. It can be cleared, which will cause the ack to be
+  // ignored.
+  private static class Receipt {
+    private final CloseableMonitor m = new CloseableMonitor();
+    private final AtomicBoolean wasAcked = new AtomicBoolean(false);
+    final Offset offset;
+
+    @GuardedBy("m.monitor")
```

Comment:
you violate this constraint on line 69 and 72. Instead, do the following:

1) hold the monitor for the full call of onAck
2) change the atomicBoolean for a normal boolean
3) change `waitUntilCommitted` to:
```
List<Receipt> receiptsCopy;
try (CloseableMonitor.Hold h = monitor.enter()) {
  receiptsCopy = ImmutableList.copyOf(receipts);
}
receiptsCopy.forEach(Receipt::clear);
try (CloseableMonitor.Hold h = monitor.enter()) {
      receipts.clear();
      acks.clear();
      committer.waitUntilEmpty();
    }
```

Reply:
Done